blocks/library/gallery/index.js

```diff
@@ -28,11 +26,25 @@ registerBlockType( 'core/gallery', {
 		images: {
 			type: 'array',
 			default: [],
-			source: query( 'div.wp-block-gallery figure.blocks-gallery-image img', {
-				url: attr( 'src' ),
-				alt: attr( 'alt' ),
-				id: attr( 'data-id' ),
-			} ),
+			source: 'query',
+			selector: 'div.wp-block-gallery figure.blocks-gallery-image img',
+			query: {
+				source: 'object',
```

Comment:
Do we need this `'object'` source at all? Is it entirely specific to the `query` behavior?

Reply:
We need something to distinguish `query( 'div', html() )` and `query( 'div', { a: html(), b: prop() }` and it's not easy to avoid ambiquity between an object defining thee shape of the sub sources or an object defining a regular source.

Reply:
Hmm, I see the issue. I don't think we want to natively support the concept of sub-sources, but rather enable specific sources like `query` to encapsulate the behavior they need. In other words, I'd really prefer we not have `object` be a proper source. Will poke at this a bit, since it's not obvious what a solution would be.

Reply:
Thinking about this more, I think this matcher can make sense even without query if you want to shape an attribute as an object composed from several sub matchers. 

Anyway, I'm open to suggestions here.

Reply:
In that case, it still doesn't seem to follow that the _source_ is an object (at least this is how it reads when seeing `source: 'object'`), but rather the desired shape is an object of a particular structure. The feature itself is not one I think we need to support at this point either.

_(Will take a closer look shortly for more concrete suggestions)_

Reply:
A few ideas:

- We could support one or the other, but not both, where query can either return an array of objects shaped of a particular structure, or an array of a particular source value
- We could split the two options by property name, e.g. `queryObject` (array of sourced object shape) and `query`
- We could test the presence of `query.source` and infer if it's defining a shape or the nested source

Reply:
> We could support one or the other, but not both,

We already need both in our blocks

> We could split the two options by property name

Maybe that's the most actionable option.

> We could test the presence of query.source 

What if `source` is a key of the subobject?

Reply:
>We already need both in our blocks

Something like `query( 'img', node() )` could be restructured as `query( 'img', { node: node() } )` (the array of object form, to normalize).

>What if `source` is a key of the subobject?

The idea is that:

- `query: { source: 'node' } == query( node() )`
- `query: { foo: { source: 'attribute' } } == query( { foo: attr( '...' ) } )`

Reply:
> What if source is a key of the subobject?

What if I want `query( { source: attr( 'src' ) } )`

>  Something like query( 'img', node() ) could be restructured as query( 'img', { node: node() } ) (the array of object form, to normalize)

I'll check if I can refactor the current usage like this.

Reply:
>What if I want `query( { source: attr( 'src' ) } )`

Yes, this is the main caveat, and I'd meant to mention it in my original comment. The simplest answer if we went this direction would be to simply not support it, though I can see how this could potentially lead to some confusion.

Reply:
I tried to always use the object notation in our current blocks but it adds a lot of burden on the `edit` function where you need to adapt the attribute shape to your usage. (mapping and transforming) which is not great.

honestly, I still think the current approach is the best approach (consistent and unambiguous), but happy to follow your preferred choice to move this PR forward.

Reply:
>honestly, I still think the current approach is the best approach (consistent and unambiguous)

Personally I find the `object` source to be very ambiguous when considered not just within the context of the `query` source.

The examples I'm seeing where we'd need to change the structure for a flattened query array are: pullquote, quote, and text columns, correct? And because of how they interact with Editable? Are these not the same types of blocks we'd be porting to use proper block nesting? Wondering if we're optimizing for a problem that won't exist.

Reply:
Correct, these are the concerned blocks. I didn't consider quotes as container for paragraph blocks but I guess you're right, it's a good fit. 

So what's the "temporary" option we should go with right now? removing the non-object `query` and transforming the value in `edit/save` for these blocks?

Reply:
>So what's the "temporary" option we should go with right now? removing the non-object `query` and transforming the value in `edit/save` for these blocks?

I'd probably be okay with this, yes. Motivation to get progress on nested blocks 😄

Reply:
updated in a4c2982

Reply:
Yes, quotes would be containers for paragraph blocks, and potentially support a few others, like lists, etc.